### PR TITLE
Replaced the child-observer with a double linked list

### PR DESCRIPTION
**Pull request by limemloh**
implemented a double linked list for the children list, to be able to add and remove children in constant time.

### CODE REVIEW

File: src/behavior.ts [hunk 1]
```diff
@@ -97,7 +94,9 @@ export abstract class Behavior<A> extends Reactive<A>
   }
   push(a: any): void {
     this.last = this.pull();
-    this.child.push(this.last);
+    for (const child of this.children) {
+      child.push(this.last);
+    }
```
Comment (paldepind) [inline src/behavior.ts]: These three lines occur a lot. Adding this as a method would avoid the repetition.

File: src/behavior.ts [hunk 2]
```diff
@@ -562,16 +575,18 @@ class ActiveScanCombineBehavior<A> extends ActiveBehavior<A> {
...
     }
   }
   pushIdx(a: A, index: number): void {
     this.last = this.accumulators[index](a, this.last);
-    if (this.child) {
-      this.child.push(this.last);
-    }
+    for (const child of this.children) {
+      child.push(this.last);
+    }
```
Comment (paldepind) [inline src/behavior.ts]: Trailing white space.

File: src/future.ts [hunk 3]
```diff
@@ -192,16 +193,17 @@ export function fromPromise<A>(p: Promise<A>): Future<A> {
  * @private
  */
 export class BehaviorFuture<A> extends SinkFuture<A> implements Observer<A> {
+  node= new Node(this);
```
Comment (paldepind) [inline src/future.ts]: Missing space before `=`. Prettier should fix that 😄

File: src/linkedlist.ts [hunk 4]
```diff
@@ -14,6 +24,60 @@ export function fromArray<A>(values: A[]): Cons<A> {
...
+  }
+}
+export class Node<A> {
+  public prev: Node<A> | undefined;
+  public next: Node<A> | undefined;
+  constructor(
+    public value: A
+  ) { }
+  static of<A>(value: A): Node<A> {
+    return new Node(value);
```
Comment (paldepind) [inline src/linkedlist.ts]: Should this method be deleted? The rest of the code uses `new Node`.

File: src/linkedlist.ts [hunk 5]
```diff
@@ -14,6 +24,60 @@ export function fromArray<A>(values: A[]): Cons<A> {
...
+      yield head.value;
+      head = head.next;
+    }
+  }
+}
+export class Node<A> {
+  public prev: Node<A> | undefined;
+  public next: Node<A> | undefined;
+  constructor(
+    public value: A
```
Comment (paldepind) [inline src/linkedlist.ts]: Are the values of `Node`s ever mutated? If not we could make this property `readonly`?

File: src/linkedlist.ts [hunk 6]
```diff
@@ -14,6 +24,60 @@ export function fromArray<A>(values: A[]): Cons<A> {
   return list;
 }
 
+/**
+ * A doubly linked list. Updates are done by mutating. Prepend, append
+ * and remove all run in O(1) time.
+ */
+export class DoubleLinkedList<A> {
```
Comment (paldepind) [inline src/linkedlist.ts]: The double linked list looks really good 👍

File: src/linkedlist.ts [hunk 7]
```diff
@@ -14,6 +24,60 @@ export function fromArray<A>(values: A[]): Cons<A> {
...
+  }
+  *[Symbol.iterator](): IterableIterator<A> {
+    let {head} = this;
+    while (head !== undefined) {
+      yield head.value;
+      head = head.next;
+    }
+  }
+}
+export class Node<A> {
```
Comment (paldepind) [inline src/linkedlist.ts]: Missing linebreak between the classes.

File: src/common.ts [hunk 8]
```diff
@@ -162,22 +107,37 @@ export abstract class Reactive<A> implements Observer<any> {
...
+    for (const parent of this.parents) {
+      const node = new Node(this);
+      this.listenerNodes = cons({node, parent}, this.listenerNodes);
+      parent.addListener(node);
+      const parentState = parent.state;
+      if (parentState !== State.Push) {
+        this.state = parentState;
+      }
+    }
+    // this.state = addListenerParents(this, this.parents, State.Push);
```
Comment (paldepind) [inline src/common.ts]: This commented line should be deleted.

File: src/behavior.ts [hunk 9]
```diff
@@ -267,6 +272,7 @@ export function ap<A, B>(
 }
 
 class ChainOuter<A> extends Behavior<A> {
+  node = new Node(this);
```
Comment (paldepind) [inline src/behavior.ts]: The `node` property isn't `private` here and in a few other places.

File: src/linkedlist.ts [hunk 10]
```diff
@@ -1,5 +1,15 @@
 export class Cons<A> {
```
Comment (paldepind) [inline src/linkedlist.ts]: Can we think of a better name for this file now that it is no longer just a linked list? `datastructures.ts`? `lists.ts`?

Comment (limemloh) [reply src/linkedlist.ts]: I agree that it should be renamed. I would prefer `datastructures.ts`

File: src/future.ts (hunk 3 above)
Comment (limemloh) [reply src/future.ts]: I'll run prettier on everything

File: src/linkedlist.ts (hunk 4 above)
Comment (limemloh) [reply src/linkedlist.ts]: yea you are right

File: src/linkedlist.ts (hunk 5 above)
Comment (limemloh) [reply src/linkedlist.ts]: it does get mutated when appending or deleting nodes in the list

File: src/behavior.ts (hunk 1 above)
Comment (limemloh) [reply src/behavior.ts]: Yes, I've considered it but did not do it, in hope that it would give better performance.

File: src/behavior.ts [hunk 11]
```diff
@@ -130,7 +127,9 @@ export abstract class ProducerBehavior<A> extends Behavior<A> {
     const changed = a !== this.last;
     this.last = a;
     if (this.state === State.Push && changed) {
-      this.child.push(a);
+      for (const child of this.children) {
+        child.push(this.last);
+      }
```
Comment (paldepind) [inline src/behavior.ts]: Can we use `pushToChildren` here?

File: src/now.ts [hunk 12]
```diff
@@ -92,15 +93,16 @@ export function perform<A>(comp: IO<A>): Now<Future<A>> {
...
-    s.addListener(this);
+    s.addListener(this.node);
     this.state = State.Push;
   }
   push(io: IO<A>): void {
     runIO(io).then((a: A) => {
-      if (this.child !== undefined) {
-        this.child.push(a);
+      for (const child of this.children) {
+        child.push(a);
```
Comment (paldepind) [inline src/now.ts]: Can we use `pushToChildren` here? `Stream` extends `Reactive` so streams should also have the `pushToChildren` method I think.

File: src/now.ts [hunk 13]
```diff
@@ -139,8 +142,8 @@ class PerformIOStreamLatest<A> extends ActiveStream<A> {
         } else {
           this.newest = time;
         }
-        if (this.child !== undefined) {
-          this.child.push(a);
+        for (const child of this.children) {
+          child.push(a);
```
Comment (paldepind) [inline src/now.ts]: `pushToChildren`

File: src/now.ts [hunk 14]
```diff
@@ -182,8 +186,8 @@ class PerformIOStreamOrdered<A> extends ActiveStream<A> {
   pushFromBuffer(): void {
     while (this.buffer[0] !== undefined) {
       const { value } = this.buffer.shift();
-      if (this.child !== undefined) {
-        this.child.push(value);
+      for (const child of this.children) {
+        child.push(value);
```
Comment (paldepind) [inline src/now.ts]: `pushToChildren`

File: src/stream.ts [hunk 15]
```diff
@@ -366,13 +374,16 @@ class SnapshotWithStream<A, B, C> extends Stream<C> {
     super();
   }
   push(a: A): void {
-    this.child.push(this.fn(a, this.behavior.at()));
+    const c = this.fn(a, this.behavior.at());
+    for (const child of this.children) {
+      child.push(c);
+    }
```
Comment (paldepind) [inline src/stream.ts]: `pushToChildren`

File: src/time.ts [hunk 16]
```diff
@@ -60,7 +65,9 @@ class DebounceStream<A> extends Stream<A> {
   push(a: A): void {
     clearTimeout(this.timer);
     this.timer = setTimeout(() => {
-      this.child.push(a);
+      for (const child of this.children) {
+        child.push(a);
+      }
```
Comment (paldepind) [inline src/time.ts]: `pushToChildren`

File: src/time.ts [hunk 17]
```diff
@@ -38,7 +41,9 @@ class ThrottleStream<A> extends Stream<A> {
   private isSilenced: boolean = false;
   push(a: A): void {
     if (!this.isSilenced) {
-      this.child.push(a);
+      for (const child of this.children) {
+        child.push(a);
+      }
```
Comment (paldepind) [inline src/time.ts]: `pushToChildren`

File: src/time.ts [hunk 18]
```diff
@@ -22,7 +21,11 @@ export class DelayStream<A> extends Stream<A> {
     return s.map(({ time, value }) => ({ time: time + this.ms, value }));
   }
   push(a: A): void {
-    setTimeout(() => this.child.push(a), this.ms);
+    setTimeout(() => {
+      for (const child of this.children) {
+        child.push(a);
+      }
```
Comment (paldepind) [inline src/time.ts]: `pushToChildren`

**Pull request merged**